annotations/src/lib.rs

```diff
@@ -22,6 +25,28 @@ macro_rules! abstract_value {
     };
 }
 
+/// An empty trait for identifying MIRAI tag types.
+pub trait MiraiTagKind {}
+
+/// A generic tag type to allow user-customizable tag kind and control mask.
+pub struct MiraiTag<KIND: MiraiTagKind, const MASK: u128> {
+    kind: std::marker::PhantomData<*const KIND>,
+}
+
+/// A trait to identify MIRAI tag types. The trait is sealed and is not meant to be implemented
+/// outside of the crate `mirai_annotations`.
+pub trait MiraiSealedTag: private::Sealed {}
```

Comment:
Instead of using the type system, in this rather convoluted way, to enforce the requirement that every tag type specifies its transfer behavior, we could do the checking in MIRAI. I.e. keep the existing design, which is that a tag can be any kind of type, but give a MIRAI diagnostic if the type does not have a generic parameter that is a constant  of type TransferMask defined as an alias for u128 in MIRAI Annotations.